File: swift/1160-e-find-words-that-can-be-form-by-characters.swift
// https://leetcode.com/problems/find-words-that-can-be-formed-by-characters/
class Solution {
    func countCharacters(_ words: [String], _ chars: String) -> Int {

        var count = Array(repeating: 0, count: 26)

        func getIndexFromChar(char: Character) -> Int {
            let charAscii = Int(char.asciiValue ?? 0)
            let aAscii = Int(Character("a").asciiValue ?? 0)
            return charAscii-aAscii
        }

        for char in chars {
            count[getIndexFromChar(char: char)] += 1
        }

        func isCanForm(word: String, count: [Int]) -> Bool {
            var runningCount = count
            for c in word {
                if runningCount[getIndexFromChar(char: c)] == 0 {
                    return false
                }
                runningCount[getIndexFromChar(char: c)] -= 1
            }

            return true
        }

        var result = 0

        for word in words {
            if isCanForm(word: word, count: count) {
                result += word.count
            }
        }
        return result
    }
}

File: python/1160-e-find-words-that-can-be-form-by-characters.py
# https://leetcode.com/problems/find-words-that-can-be-formed-by-characters/
from typing import List
from collections import defaultdict

class Solution:
    def countCharacters(self, words: List[str], chars: str) -> int:

        countChars = [0 for i in range(26)]

        for c in chars:
            countChars[ord(c) - ord('a')] += 1

        def isCanForm(word):
            chars = countChars.copy()
            for c in word:
                index = ord(c) - ord('a')
                if chars[index] == 0:
                    return False
                chars[index] -= 1
            
            return True
        
        count = 0

        for word in words:
            isCan = isCanForm(word)
            if isCan:
                count += len(word)

        return count

# Test runner
if __name__ == "__main__":
    solution = Solution()
    
    # Test cases
    test_cases = [
        (["cat","bt","hat","tree"], "atach", 6),           # "cat" and "hat" can be formed
        (["hello","world","leetcode"], "welldonehoneyr", 10), # "hello" and "world" can be formed
        (["a"], "a", 1),                                   # Single character word and chars
        (["aa"], "a", 0),                                  # Not enough characters
        (["word"], "world", 4),                            # Single word that can be formed
        (["none"], "xyz", 0),                              # No words can be formed
        (["good","bad","test"], "goodbad", 7),            # Multiple words can be formed
        (["", "test"], "test", 4),                        # Empty string in words
        (["aaa","bbb"], "aaabbb", 6),                     # Multiple words with repeated chars
    ]
    
    # Run tests
    for i, (words, chars, expected) in enumerate(test_cases, 1):
        result = solution.countCharacters(words, chars)
        print(f"Test {i}:")
        print(f"Words: {words}")
        print(f"Characters: {chars}")
        print(f"Expected: {expected}")
        print(f"Result: {result}")
        print(f"{'✓ Passed' if result == expected else '✗ Failed'}")
        print()